### DISCUSSION TITLE
fastled problem

### DISCUSSION THREAD

**Discussion by salvodjkr** [tttapa/Control-Surface#1008] [General]
![IMG_1338](https://github.com/tttapa/Control-Surface/assets/66134251/e9f92d20-5605-4eac-9ad0-0e37a8752c42)

hi piter, congratulations for this wonderful project. I'm making my first ssl style midi controller. I have no problems with the potentiometers and buttons but I can't get fastled to work. I tried the fastled example to make the led flash and it works but when I use the example you gave in the control surface examples, it doesn't work. I would like that when I disable/enable the plugin with Ableton the LED turns off and on.

this is my main code. sorry for my English

**Comment by salvodjkr** [comment]
```
#include <Control_Surface.h>

#include <AH/Hardware/MCP23017Encoders.hpp>

#include <Wire.h>



// Type for the MCP23017 encoders (translates encoder pulses to position)
using WireType = decltype(Wire);     // The type of I²C driver to use
using EncoderPositionType = int32_t; // The type for saving encoder positions
constexpr bool IntSafe = true;       // Make it safe to call `update` in an ISR
using MCPEncoderType = MCP23017Encoders<WireType, EncoderPositionType, IntSafe>;

// Type for the MIDI encoders (translates position to MIDI messages)
struct CCMCPEncoder :GenericMIDIAbsoluteEncoder<MCPEncoderType::MCP23017Encoder,
                                               ContinuousCCSender> {
  CCMCPEncoder(MCPEncoderType::MCP23017Encoder enc, MIDIAddress address,
               int16_t multiplier = 1, uint8_t pulsesPerStep = 4)
    : GenericMIDIAbsoluteEncoder(std::move(enc), address, multiplier,
                               pulsesPerStep, {}) {}
                               
};




USBMIDI_Interface midi;
//USBMIDI_Interface usbmidi;





//constexpr pin_t ledPin = 7;


CD74HC4051 mux {
  A0,       // Analog input pin
  {3, 4, 5} // Address pins S0, S1, S2
};

// Create an array of CCPotentiometer objects that send out MIDI Control Change 
// messages when you turn the potentiometers connected to the 8 inputs of the mux.

using namespace MIDI_Notes;

//CCButton pulsanti[] { //configurazione come pulsante switch
CCButtonLatched pulsanti[] { //configurazione come pulsante toggle
  { mux.pin(0), {104, CHANNEL_1},127},
  { mux.pin(1), { MIDI_CC::General_Purpose_Controller_2, Channel_2 } },
  { mux.pin(2), { MIDI_Notes::E(4), CHANNEL_1 } },    // GEN2
  { mux.pin(3), { MIDI_CC::Channel_Volume, Channel_4 } },
};





const uint8_t interrupt_pin = 12;

// Create an object that manages the 8 encoders connected to the MCP23017.
MCPEncoderType enc {Wire, 0x20, interrupt_pin};
//                  │     │    └─ Interrupt pin
//                  │     └────── Address offset
//                  └──────────── I²C interface

// Instantiate 8 MIDI rotary encoders.
CCMCPEncoder ccencoders[] {
  {enc[0], {1, CHANNEL_16},1,4},
  {enc[1], {2, CHANNEL_16},1,4},
  {enc[2], {3, CHANNEL_16},1,4},
  {enc[3], {4, CHANNEL_16},1,4},
  {enc[4], {5, CHANNEL_16},1,4},
  {enc[5], {6, CHANNEL_16},1,4},
  {enc[6], {7, CHANNEL_16},1,4},
  {enc[7], {8, CHANNEL_16},1,4},
  };

void isr() {
  enc.update(); // Read the state of the encoders and update the positions
}

void setup() {
  //pinMode(ledPin, OUTPUT);
  //digitalWrite(ledPin, HIGH);   // turn the LED on (HIGH is the voltage level)
  Control_Surface.begin();
  
  Wire.begin(); // Must be called before enc.begin()
  Wire.setClock(800000);
  enc.begin(); // Initialize the MCP23017
  mux.begin();
  attachInterrupt(digitalPinToInterrupt(interrupt_pin), isr, LOW);
}

void loop() {
  Control_Surface.loop();
  
}
```

**Comment by salvodjkr** [comment]
I forgot to tell you that I use "remotifi" software to map ableton macros. it works very well.

**Comment by tttapa** [comment]
Which example did you try? And what board are you using?

**Comment by salvodjkr** [comment]
Thanks to the immediate response. I'm starting from scratch. I tested this code. How do I adapt this code to what I would like to do? in ableton when I use the first macro it turns the plugin on and off and this works well also thanks to the script created with remotify. my request is that when the macro is 0 the first LED of the ws2812 strip is off, however when it is a value set by me it turns on. All this to have bright feedback when the plugin is active.

```
#include <FastLED.h>
// Must be before Control Surface to enable FastLED features of Control Surface
#include <Control_Surface.h>

// Define the array of leds.
Array<CRGB, 8> leds {};
// The data pin with the strip connected.
constexpr uint8_t ledpin = 2;

USBMIDI_Interface midi;

// Create a MIDI input element that listens to all notes in the range C4 - G4
// (the range starts at C4 and has a length equal to `leds.length` == 8).
NoteRangeFastLED<leds.length> midiled {leds, MIDI_Notes::C[4]};

void setup() {
  // See FastLED examples and documentation for more information.
  FastLED.addLeds<NEOPIXEL, ledpin>(leds.data, leds.length);
  FastLED.setCorrection(TypicalPixelString);
  midiled.setBrightness(128);
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
  if (midiled.getDirty()) { // If the colors changed
    FastLED.show();         // Update the LEDs with the new colors
    midiled.clearDirty();   // Clear the dirty flag
  }
}
```

**Reply by tttapa** [reply]
Please try the following:
- Replace `USBMIDI_Interface` by `USBDebugMIDI_Interface`.
- Upload and open the Serial Monitor.
- Send a Note On message addressed to the first LED in the strip: `90 3C 7F` <kbd>Enter</kbd>
- The first LED should turn on in a yellowish color.

If this works, but not with Ableton, that means that Ableton does not send the messages that your Arduino code is expecting. You can use a MIDI monitor (or use one of the MIDI monitor examples of the library) to see what messages Ableton is sending.

**Comment by salvodjkr** [comment]
use teensy 4.1

**Comment by salvodjkr** [comment]
here's what you get with the midi input example.

<img width="906" alt="Screenshot 2024-02-20 alle 13 45 46" src="https://github.com/tttapa/Control-Surface/assets/66134251/2f68394a-f357-4527-8c09-1bb9e1b9c3cf">

in fact when I use macros it receives cc messages.

<img width="593" alt="Screenshot 2024-02-20 alle 13 46 43" src="https://github.com/tttapa/Control-Surface/assets/66134251/56bf62ce-6ef7-462d-bd7f-3e1fe7fe2224">

How can I integrate the code I wrote to you yesterday for turning on the LED using the Ableton macros?
I also tried using the shiftregister but with teensy 4.1 it doesn't work.

**Reply by tttapa** [reply]
It looks like Ableton is sending MIDI Control Change messages, so it's not too surprising that the NoteLED class and similar won't respond to them. You'll need the CCLED class, or a customization thereof.

Step one will be to figure out what the values in the “data” column mean. Consult the manual for the plugin, or reverse engineer it with trial and error.
Then have a look at this example: https://tttapa.github.io/Control-Surface-doc/Doxygen/dd/d9d/Custom-MIDI-Input-Element_8ino-example.html
You'll want to use `MIDIMessageType::ControlChange` instead of `MIDIMessageType::NoteOn`.

**Comment by benwadub** [comment]
hello @salvodjkr i also use remotely and try to output my encoders data from ableton to my custom midi controller, can you explain how you were able to output macro midi data to the teensy please? I can t figure how to route midi from output from remotely.
